accepted/00000-replace-salt-toaster.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: Replace Salt Toaster
+- Start Date: 2022-08-03
+
+# Summary
+[summary]: #summary
+
+Replace Salt Toaster with an approach that's closer to upstream's testing efforts.
+
+# Motivation
+[motivation]: #motivation
+
+<!-- - Why are we doing this? -->
+<!-- - What use cases does it support? -->
+<!-- - What is the expected outcome? -->
+
+<!-- Describe the problem you are trying to solve, and its constraints, without coupling them too closely to the solution you have in mind. If this RFC is not accepted, the motivation can be used to develop alternative solutions. -->
+
+Salt Toaster was developed a few years ago and filled a gap in the Salt world.
+Over the last years, maintaining the downstream `suse.tests` was too much effort
+-- so much that we didn't do it. Maintaining the docker images that Salt Toaster
+uses to run the upstream test suite against our patched Salt grew into a bigger
+and bigger task. We didn't keep up with upstream's changes to the test
+dependencies well and use a number of workarounds to still run a test suite.
+
+At the same time, Salt upstream improved their test suite a lot. They added new
+test types and now have a similar test framework to Salt Toaster. 
+
+The purpose of this change is to lessen our maintenance burden and to be good
+open-source citizens that collaborate with our upstream project when it comes to
+testing Salt. Additionally, we have knowledge about running a test suite within
+our team that we want to take advantage of.
+
+# Detailed design
+[design]: #detailed-design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the product to understand, and for somebody familiar with the internals to implement. -->
+
+<!-- This section should cover architecture aspects and the rationale behind disruptive technical decisions (when applicable), as well as corner-cases and warnings. Whenever the new feature creates new user interactions, this section should include examples of how the feature will be used. -->
+
+## Overview
+
+The Salt upstream test suite is executed in virtual machines (VMs) that run the
+different operating systems we support. These VMs are set up with sumaform and
+salt-jenkins. Terracumber triggers terraform to create the VMs from within a
+Jenkins job and launches the test suite, connecting to each VM over SSH.
+
+## Provisioning Test VMs with Sumaform + Salt-Jenkins
+
+A new module "salt\_testenv" is added to sumaform. This module defines a new
+role, "salt\_testenv", and the default values needed to run all test suites, e.g.
+4GiB of RAM. Sumaform uses this role to apply the `golden-images-provision.sls`
+state file from the Salt-Jenkins repository.
```

Comment:
I guess it would be great to have a generic sumaform option to run arbitrary Salt files at the end of provisioning.

Reply:
I am against this. Sumaform has a defined scope and if we generify everything we run into a kraken of features again. The slightest hickup in the wrong step causes a complete destroy and apply cycle. Let's not do more then needed. But that is only my view on things.

Reply:
A generic sumaform option could be useful if we wanted to define states to apply to a module in `main.tf`, without having to touch `salt/top.sls`. This would allow us to extend an existing "role" with arbitrary state files (that are already present in `salt/` for a single "module".

If we had that option, we could define this salt_testenv module completely in a `main.tf`. I prefer not to go this route because it would make setting up a VM harder. The proposed module defined in "modules/salt\_testenv" makes it easy to deploy such a VM without having to know the name of the state file, or the required memory.

Generally this is could be a nice feature. I believe it's orthogonal to adding this new module though.

Reply:
> I am against this. Sumaform has a defined scope and if we generify everything we run into a kraken of features again. The slightest hickup in the wrong step causes a complete destroy and apply cycle. Let's not do more then needed. But that is only my view on things.

I respect that.

> I believe it's orthogonal to adding this new module though.

Completely agreed.

FMPOV, in general, having clear mechanisms to extend functionality of a project helps in use cases in which it does not make sense to add overspecific code - although it's of course delicate to set the interface well. I will let sumaform maintainers make the final call.